README.md
# Tile38

This package contains convenience functions which map responses from tile38 to objects.

This package maps the `z` coordinate to the key `timestamp`.

## Peer Dependencies

- [Redix](https://github.com/whatyouhide/redix).

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `tile38_client` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:ex_tile38, ">= 0.0.0"},
    {:redix, ">= 0.0.0"}
  ]
end
```

## Usage

### Start link to Redix

Notice that we give this connection the name `tile38`.

```
{:ok, _} = Redix.start_link("redis://localhost:9851/", name: :tile38)
```

See documentation on usage at [https://hexdocs.pm/ex_tile38](https://hexdocs.pm/ex_tile38).



Changelog
0.3.0 Error bubbling
0.2.0 Add strong typing with %Tile38.Point{}
0.1.0 First implementation


test/tile38_test.exs
defmodule Tile38Test do
  use ExUnit.Case
  doctest Tile38
  import Tile38

  setup_all do
    {:ok, _} = Redix.start_link("redis://localhost:9851/", name: :tile38)

    :ok
  end

  setup do
    clear_database()

    :ok
  end

  describe "#f38" do
    test "request an object with fields from Tile38" do
      id = "123"

      t38("set mycollection #{id} field firstfield 10 field secondfield 20 point 10 -10 123")

      object_with_fields = f38("get mycollection #{id} withfields")

      assert(object_with_fields.fields.firstfield == "10")
      assert(object_with_fields.fields.secondfield == "20")
      assert(object_with_fields.coordinates.lat == 10)
      assert(object_with_fields.coordinates.lng == -10)
      assert(object_with_fields.coordinates.timestamp == 123)
      assert(object_with_fields.coordinates.type == "Point")
    end

    test "flip field order" do
      id = "123"
      t38("set participants #{id} field secondfield 20 field firstfield 10 point 10 -10")

      object_with_fields = f38("get participants #{id} withfields")

      assert(object_with_fields.fields.firstfield == "10")
      assert(object_with_fields.fields.secondfield == "20")
      assert(object_with_fields.coordinates.lat == 10)
      assert(object_with_fields.coordinates.lng == -10)
      assert(object_with_fields.coordinates.timestamp == nil)
      assert(object_with_fields.coordinates.type == "Point")
    end

    test "no fields, no timestamp" do
      id = "123"
      t38("set participants #{id} point 10 -10")

      result = f38("get participants #{id} withfields")

      assert(result.coordinates.lat == 10)
      assert(result.coordinates.lng == -10)
      assert(result.coordinates.timestamp == nil)
    end

    test "no fields but timestamp" do
      id = "123"
      t38("set participants #{id} point 10 -10 123")

      result = f38("get participants #{id} withfields")

      assert(result.coordinates.lat == 10)
      assert(result.coordinates.lng == -10)
      assert(result.coordinates.timestamp == 123)
    end

    test "no results" do
      id = "123"

      result = f38("get participants #{id} withfields")

      assert(result == [])
    end
  end

  describe "#n38" do
    test "no points nearby" do
      response = n38("nearby participants point 0 0")

      assert(response == [])
    end

    test "single point nearby" do
      id = "123"
      t38("set participants #{id} field my_field 123 point 10 -10 99")

      [response] = n38("nearby participants point 0 0")

      assert(response.id == "123")
      assert(response.coordinates.lat == 10)
      assert(response.coordinates.lng == -10)
      assert(response.coordinates.timestamp == 99)
    end

    test "multiple points nearby" do
      id = "0123"
      other_id = "0321"
      t38("set participants #{id} field my_field 123 field my_other_field 321 point 0 0")
      t38("set participants #{other_id} point 10 -10 10000")

      response = n38("nearby participants point 0 0")

      [first, second] = response

      assert(first.id == "0321")
      assert(first.coordinates.lat == 10)
      assert(first.coordinates.lng == -10)
      assert(first.coordinates.timestamp == 10000)
      assert(second.id == "0123")
      assert(second.coordinates.lat == 0)
      assert(second.coordinates.lng == 0)
      assert(second.coordinates.timestamp == nil)
      assert(second.fields.my_field == "123")
      assert(second.fields.my_other_field == "321")
    end
  end
end


mix.exs
defmodule Tile38.MixProject do
  use Mix.Project

  def project do
    [
      app: :ex_tile38,
      version: "0.3.0",
      elixir: "~> 1.11",
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description:
        "Elixir wrapper for Tile38 client. Formats responses to common queries for convenience.",
      package: package()
    ]
  end

  defp package do
    [
      maintainers: ["lpender"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/lunchtime-labs/ex_tile38"}
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:ex_doc, ">= 0.0.0", only: :dev},
      {:jason, "~> 1.0"},
      {:redix, ">= 0.0.0"}
    ]
  end
end


lib/tile38.ex
defmodule Tile38 do
  @moduledoc """
  Elixir wrapper for [Tile38](https://tile38.com/) client. Formats responses to common queries for convenience.

  ## Getting started

  This package requires [Redix](https://github.com/whatyouhide/redix).

  ```
  # mix.exs
    defp deps do
      {:ex_tile38, ">= 0.0.0"},
      {:redix, ">= 0.0.0"}
    end
  ```

  Before calling any commands, you must initialize a named Redix connection, i.e.

  ```
  {:ok, _} = Redix.start_link("redis://localhost:9851/", name: :tile38)
  ```

  If you want to use a name other than `:tile38`, you can pass the name as an atom
  to any method as the second argument.
  """

  @default_redix_key :tile38

  @doc """
    Clear the entire Tile38 database.
  """
  def clear_database(redix_key \\ @default_redix_key) do
    t38("flushdb", redix_key)
  end

  @doc """
  Send a `get` command `withfields` to Tile38 and and return the response as JSON.

  ## Parameters

    - command: String of the exact command you wish to send to Tile38.
    - redix_key: (Optional) Atom with the name of the redix connection. Defaults to `:tile38`.

  ## Examples

      iex> Tile38.t38("set mycollection my_id field firstfield 10 field secondfield 20 point 10 -10 123")
      iex> Tile38.f38("get mycollection my_id withfields")
      %Tile38.Point{
        fields: %{
          firstfield: "10",
          secondfield: "20"
        },
        coordinates: %{
          lat: 10,
          lng: -10,
          timestamp: 123,
          type: "Point"
        }
      }
  """
  def f38(command, redix_key \\ @default_redix_key) do
    resp = t38(command, redix_key)

    case resp do
      nil ->
        []

      [point, fields] ->
        %{type: type, coordinates: [lng, lat | timestamp]} = Jason.decode!(point, keys: :atoms)

        coordinates = %{type: type, lat: lat, lng: lng, timestamp: List.first(timestamp)}

        fields = list_to_map(fields, %{})

        %Tile38.Point{
          coordinates: coordinates,
          fields: fields
        }

      [point] ->
        %{type: type, coordinates: [lng, lat | timestamp]} = Jason.decode!(point, keys: :atoms)

        coordinates = %{type: type, lat: lat, lng: lng, timestamp: List.first(timestamp)}

        %Tile38.Point{
          coordinates: coordinates
        }

      {:error, err} ->
        {:error, err}
    end
  end

  @doc """
  Send a `jget` command to Tile38 and parse the response as JSON.

  ## Parameters

    - command: String of the exact command you wish to send to Tile38.
    - redix_key: (Optional) Atom with the name of the redix connection. Defaults to `:tile38`.

  ## Examples

      iex> Tile38.t38("jset user 901 name.first Tom")
      iex> Tile38.j38("jget user 901")
      %{name: %{first: "Tom"}}
  """
  def j38(command, redix_key \\ @default_redix_key) do
    resp = t38(command, redix_key)

    case resp do
      {:error, err} ->
        {:error, err}

      nil ->
        nil

      record ->
        Jason.decode!(record, keys: :atoms)
    end
  end

  @doc """
  Send a `nearby` or `scan` command to Tile38 and and return the response as JSON.

  ## Parameters

    - command: String of the exact command you wish to send to Tile38.
    - redix_key: (Optional) Atom with the name of the redix connection. Defaults to `:tile38`.

  ## Examples

      iex> t38("set mycollection my_id field firstfield 10 point 10 -10 1000")
      iex> n38("nearby mycollection point 10 -10")
      [
        %Tile38.Point{
          id: "my_id",
          coordinates: %{
            lat: 10,
            lng: -10,
            timestamp: 1000,
            type: "Point",
          },
          fields: %{
            firstfield: "10"
          }
        }
      ]
  """
  def n38(command, redix_key \\ @default_redix_key) do
    resp = t38(command, redix_key)

    case resp do
      [_, []] ->
        []

      [_, data] ->
        parse_nearby(data, [])

      {:error, err} ->
        {:error, err}
    end
  end

  @doc """
  Send any command to Tile38 and return the response with no special formatting applied.

  ## Parameters

    - command: String of the exact command you wish to send to Tile38.
    - redix_key: (Optional) Atom with the name of the redix connection. Defaults to `:tile38`.

  ## Examples

      iex> Tile38.t38("set mycollection my_id field firstfield 10 point 10 -10 1000")
      iex> Tile38.t38("get mycollection my_id withfields")
      [~s/{"type":"Point","coordinates":[-10,10,1000]}/, ["firstfield", "10"] ]
  """
  def t38(command, redix_key \\ @default_redix_key) do
    case Redix.command(redix_key, ~w(#{command})) do
      {:ok, resp} ->
        resp

      {:error, err} ->
        {:error, err}
    end
  end

  # Private methods

  defp list_to_map([], acc) do
    acc
  end

  defp list_to_map([key, value | tail], acc) do
    acc = Map.put(acc, String.to_atom(key), value)

    list_to_map(tail, acc)
  end

  defp parse_nearby([], acc) do
    acc
  end

  defp parse_nearby([[id, point | fields] | rest], acc) do
    point = Jason.decode!(point, keys: :atoms)

    fields =
      case is_list(fields) do
        true ->
          List.first(
            Enum.map(fields, fn field ->
              list_to_map(field, %{})
            end)
          )

        false ->
          %{}
      end

    %{coordinates: [lng, lat | timestamp], type: type} = point

    object = %Tile38.Point{
      id: id,
      coordinates: %{
        lat: lat,
        lng: lng,
        timestamp: List.first(timestamp),
        type: type
      },
      fields: fields
    }

    acc = [object | acc]

    parse_nearby(rest, acc)
  end
end
